codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/EndpointPlugin.kt

```diff
@@ -0,0 +1,35 @@
+package software.amazon.smithy.aws.swift.codegen.plugins
+
+import software.amazon.smithy.aws.swift.codegen.AWSServiceTypes
+import software.amazon.smithy.aws.swift.codegen.AWSSwiftDependency
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.swift.codegen.SwiftWriter
+import software.amazon.smithy.swift.codegen.integration.Plugin
+import software.amazon.smithy.swift.codegen.integration.ServiceConfig
+import software.amazon.smithy.swift.codegen.model.buildSymbol
+
+class EndpointPlugin(private val serviceConfig: ServiceConfig) : Plugin {
+    override val className: Symbol
+        get() = buildSymbol {
+            this.name = "EndpointPlugin"
+            this.namespace = AWSSwiftDependency.AWS_CLIENT_RUNTIME.target
+            dependency(AWSSwiftDependency.AWS_CLIENT_RUNTIME)
+        }
+
+    override val isDefault: Boolean
+        get() = false
+
+    override fun render(writer: SwiftWriter) {
+        writer.openBlock("public class EndpointPlugin: Plugin {", "}") {
+            writer.write("private var endpointResolver: \$L", AWSServiceTypes.EndpointResolver)
+            writer.openBlock("public init(endpointResolver: \$L) {", "}", AWSServiceTypes.EndpointResolver) {
+                writer.write("self.endpointResolver = endpointResolver")
+            }
+            writer.openBlock("public func configureClient(clientConfiguration: ClientRuntime.ClientConfiguration) throws {", "}") {
+                writer.openBlock("if var config = clientConfiguration as? ${serviceConfig.typeName} {", "}") {
+                    writer.write("config.endpointResolver = self.endpointResolver")
+                }
+            }
+        }
+    }
```

Comment:
Why does the endpoint resolver plugin have to be code-generated? Can we maybe add the property `endpointResolver` to the base client config protocol in smithy-swift and have a plugin written in AWSClientRuntime that works the same as other plugins?

Reply:
Oh, so the base protocol itself is code-generated for each service. Maybe we can delete the codegen for that protocol given it seems to be same thing for every service, move it to AWSClientRuntime, and do the approach I mentioned in original comment. That way we can reduce 5k lines of fundamentally duplicate codegen.

Reply:
Hmm.. forgot that endpoint params for each service has different values, and maybe that makes the approach I suggested be not feasible.